src/renderer/builder.rs

```diff
@@ -0,0 +1,287 @@
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_style_template(mut self, style_template: impl ToString) -> Self {
+        self.style_template = style_template.to_string();
+        self
+    }
+
+    /// Sets sprite URL template
+    ///
+    /// Default: "/{path}/sprite{scale}.{format}"
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_sprites_template(mut self, sprites_template: impl ToString) -> Self {
+        self.sprites_template = sprites_template.to_string();
+        self
+    }
+
+    /// Sets glyph URL template
+    ///
+    /// Default: "/font/{fontstack}/{start}-{end}.pbf"
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_glyphs_template(mut self, glyphs_template: impl ToString) -> Self {
+        self.glyphs_template = glyphs_template.to_string();
+        self
+    }
+
+    /// Sets tile URL template
+    ///
+    /// Default: "/{path}"
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_tile_template(mut self, tile_template: impl ToString) -> Self {
+        self.tile_template = tile_template.to_string();
+        self
+    }
+
+    /// Sets API key parameter name
+    ///
+    /// Default: ""
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_api_key_parameter_name(mut self, api_key_parameter_name: impl ToString) -> Self {
+        self.api_key_parameter_name = api_key_parameter_name.to_string();
+        self
+    }
+
+    /// Sets API key
+    ///
+    /// Default: ""
+    #[must_use]
+    #[allow(clippy::needless_pass_by_value, reason = "false positive")]
+    pub fn with_api_key(mut self, api_key: impl ToString) -> Self {
+        self.api_key = api_key.to_string();
+        self
+    }
+
+    /// Sets whether API key is required
+    ///
+    /// Default: `false`
+    #[must_use]
+    pub fn set_requires_api_key(mut self, requires_api_key: impl Into<bool>) -> Self {
+        self.requires_api_key = requires_api_key.into();
+        self
+    }
+
+    /// Builds a static image renderer
+    #[must_use]
+    pub fn build_static_renderer(self) -> ImageRenderer<Static> {
+        // TODO: Should the width/height be passed in here, or have another `build_static_with_size` method?
+        ImageRenderer::new(MapMode::Static, &self)
+    }
+
+    /// Builds a tile renderer
+    #[must_use]
+    pub fn build_tile_renderer(self) -> ImageRenderer<Tile> {
+        // TODO: Is the width/height used for this mode?
+        ImageRenderer::new(MapMode::Tile, &self)
+    }
+}
+
+impl<S> ImageRenderer<S> {
+    /// Creates a new renderer instance
+    fn new(map_mode: MapMode, opts: &ImageRendererBuilder) -> Self {
+        let cache = opts
+            .cache_path
+            .as_deref()
+            .map(|path| path.display().to_string())
```

Comment:
you should never use `display` for non-display usecases. If its not a valid UTF-8, fail.  In the future, on windows, we could modify MLN to allow those strings if we really want to

Reply:
Seems kind of difficult to do, lets see what I can do
- https://github.com/dtolnay/cxx/issues/1291